lib/menubar/utils.lua

```diff
@@ -243,17 +246,49 @@ end
 
 --- Parse a directory with .desktop files recursively.
 -- @tparam string dir The directory.
--- @treturn table Paths of found .desktop files.
-function utils.parse_dir(dir)
-    local programs = {}
-    local files = io.popen('find '.. dir .." -name '*.desktop' 2>/dev/null")
-    for file in files:lines() do
-        local program = utils.parse_desktop_file(file)
-        if program then
-            table.insert(programs, program)
+-- @tparam function callback Will be fired when all the files will be parsed.
+-- @tparam table callback.programs Paths of found .desktop files.
+function utils.parse_dir(dir, callback)
+    gio.Async.start(function()
+        local f = gio.File.new_for_path(dir)
+        -- Except for "NONE" there is also NOFOLLOW_SYMLINKS
+        local enum, err = f:async_enumerate_children("standard::name", gio.FileQueryInfoFlags.NONE)
+        if not enum then
+            debug.print_error(err)
+            return callback(nil)
         end
-    end
-    return programs
+        local subdirs_to_parse = 0
+        local files_per_call = 100 -- Actual value is not that important
+        local programs = {}
+        while true do
+            local list, enum_err = enum:async_next_files(files_per_call, glib.PRIORITY_DEFAULT)
+            if enum_err then
+                debug.print_error(enum_err)
+                return callback(nil)
+            end
+            for _, info in ipairs(list) do
+                local file_type = info:get_file_type()
+                local file_path = f:get_child(info:get_name()):get_path()
+                if file_type == 'REGULAR' then
+                    local program = utils.parse_desktop_file(file_path)
+                    if program then
+                        table.insert(programs, program)
+                    end
+                elseif file_type == 'DIRECTORY' then
+                    subdirs_to_parse = subdirs_to_parse + 1
+                    utils.parse_dir(file_path, function(more_programs)
+                        subdirs_to_parse = subdirs_to_parse - 1
+                        if more_programs then
+                            awful_util.table.merge(programs, more_programs)
+                        end
+                    end)
+                end
+            end
+            if #list == 0 and subdirs_to_parse == 0 then
+                return callback(programs)
+            end
+        end
```

Comment:
Please add `enum:async_close()` here (after the `while true`-loop). When writing the code yesterday I thought "the garbage collector will clean up". However, this file enumeration keeps a file descriptor open and there is an upper limit on the number of file descriptors that can be open. Thus, this should be closed when done.